### DISCUSSION TITLE
Backend tests fail if copier used to clone

### DISCUSSION THREAD

**Discussion by swelborn** [fastapi/full-stack-fastapi-template#1135] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched in the documentation/README.
- [X] I already searched in Google "How to do X" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs/README and didn't find an answer.

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```Text
N/A
```

### Description

Used copier to template the repo. When I do this, the users tests fail. 

The `.env` has this for default:
```
SECRET_KEY=changethis
FIRST_SUPERUSER=admin@example.com
FIRST_SUPERUSER_PASSWORD=changethis
```

mine was changed to this:
```
SECRET_KEY=changeme
FIRST_SUPERUSER=swelborn@lbl.gov
FIRST_SUPERUSER_PASSWORD=HMpzt9&VAgao6%
```
And I get the following errors:
```
ERROR app/tests/api/routes/test_users.py::test_get_users_superuser_me - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_create_user_new_email - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_get_existing_user - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_create_user_existing_username - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_retrieve_users - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_password_me - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_password_me_incorrect_password - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_password_me_same_password_error - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_user - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_user_not_exists - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_update_user_email_exists - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_delete_user_super_user - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_delete_user_not_found - KeyError: 'access_token'
ERROR app/tests/api/routes/test_users.py::test_delete_user_current_super_user_error - KeyError: 'access_token'
```
When I revert back to the default values above, they all pass.

### Operating System

macOS

### Operating System Details

I am using docker compose for this.

### Python Version

3.12

**Comment by swelborn** [comment]
When I have it on the default settings, I get the following:

```
app/tests/api/routes/test_items.py {'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTIyNzAzMTQsInN1YiI6IjIyNSJ9.vt3jx5s0fGr476fvBXk_X1DjABjTKT7Xav0DTrh8ldI'}
...........
app/tests/api/routes/test_login.py ..{'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTIyNzAzMTgsInN1YiI6IjIyNSJ9.RVKQHTfd9jnpW9tfbXyGuzcAdSL2475rgEfrNLARojo'}
.....
app/tests/api/routes/test_users.py {'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTIyNzAzMTksInN1YiI6IjIyNSJ9.NHLF_j1VruwnlWiYOb5e9Nf_3NfTt7Q_TzNx3HyREWc'}
.........................
```

I added a print statement here:

```python
@pytest.fixture(scope="module")
def superuser_token_headers(client: TestClient) -> dict[str, str]:
    print(get_superuser_token_headers(client))
    return get_superuser_token_headers(client)
```

and added `-s` to the test script here:
```
coverage run --source=app -m pytest -s
```

After changing password to `FIRST_SUPERUSER_PASSWORD=HELLO` the `{'Authorization':...` no longer shows up for `test_users.py`, but does on `test_items` and `test_login`:

```
app/tests/api/routes/test_items.py {'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTIyNzA1MTYsInN1YiI6IjI2MSJ9.ETxxIi_ajtc0tmwK7La7_O_iGnpV3rRDUiBI9npw9kk'}
...........
app/tests/api/routes/test_login.py ..{'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTIyNzA1MTksInN1YiI6IjI2MSJ9.GZSXqxs6JCEsfDon_5xayr5oyYnk3tjELFlxKhakF_c'}
.....
app/tests/api/routes/test_users.py E.EE..E.E.EE.E...EEEE.EE.

```

**Comment by swelborn** [comment]
Appears that this is the problem:

```python
def test_reset_password(
    client: TestClient, superuser_token_headers: dict[str, str]
) -> None:
    token = generate_password_reset_token(email=settings.FIRST_SUPERUSER)
    data = {"new_password": "changethis", "token": token}
    r = client.post(
        f"{settings.API_V1_STR}/reset-password/",
        headers=superuser_token_headers,
        json=data,
    )
    assert r.status_code == 200
    assert r.json() == {"message": "Password updated successfully"}
```

If you change this to :

```python
def test_reset_password(
    client: TestClient, superuser_token_headers: dict[str, str]
) -> None:
    token = generate_password_reset_token(email=settings.FIRST_SUPERUSER)
    data = {"new_password": settings.FIRST_SUPERUSER_PASSWORD, "token": token}
    r = client.post(
        f"{settings.API_V1_STR}/reset-password/",
        headers=superuser_token_headers,
        json=data,
    )
    assert r.status_code == 200
    assert r.json() == {"message": "Password updated successfully"}
```

all tests succeed.

**Reply by t11c** [reply]
Thanks! this solved my problem. I spent several hours yesterday to investigate this error. But did you understand the reasone for this error/solution

**Reply by t11c** [reply]
I just tested somethin else; changing the scope of the pytest fixture `db` in `app/tests/conftest.py` from `session` to `module` also solves this issue. Without having to change the value for `new_password` as above

from this:

```python
@pytest.fixture(scope="session", autouse=True)
def db() -> Generator[Session, None, None]:
    with Session(engine) as session:
        init_db(session)
        yield session
        statement = delete(Item)
        session.execute(statement)
        statement = delete(User)
        session.execute(statement)
        session.commit()
```

into this:

```python
@pytest.fixture(scope="module", autouse=True)
def db() -> Generator[Session, None, None]:
    with Session(engine) as session:
        init_db(session)
        yield session
        statement = delete(Item)
        session.execute(statement)
        statement = delete(User)
        session.execute(statement)
        session.commit()
```

**Reply by swelborn** [reply]
Yeah I think the reset password test changes the password for any future tests. Your solution may be better than this, because I think it will set up a db for each test module. But I'm not sure what the desired behavior is, so we should wait for @tiangolo or someone more familiar before PR.

**Reply by callmeBalloch** [reply]
I had the same problem! Ty guys! This should be updated in the repo tho

**Comment by swelborn** [comment]
Added PR here: https://github.com/tiangolo/full-stack-fastapi-template/pull/1213

**Comment by verbal** [comment]
I also ran into this. I'm not sure why this test is written to break if you actually change the FIRST_SUPERUSER_PASSWORD. If instead of using "changethis" you use the FIRST_SUPERUSER_PASSWORD, whatever its set to, then its not a great test because you don't actually test if the password got changed since you're changing it to the same password. In my opinion there are two approaches, 1) you can create a throw away user (like test_create_user_by_normal_user) to actually change the password for, then it won't mess with the superuser's password, or 2) once you change it to "changethis", change it back to the original password, thus restoring the state (this is what test_update_password_me does). 

Is the idea to only run the tests with totally clean data and have it return to that state after the tests?

**Comment by YuriiMotov** [comment]
Fixed by https://github.com/fastapi/full-stack-fastapi-template/pull/1499
